From beef2a688e4496fd47dfcd75e8f084c2b06409ab Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Thu, 30 Jul 2015 12:28:31 +0900
Subject: [PATCH 2/2] libv4l2: Support mmap() to libv4l-plugin

This support adds the call to mmap() from libv4l-plugin
prior to the system call.
---
 lib/include/libv4l-plugin.h | 3 ++-
 lib/libv4l2/libv4l2.c       | 8 ++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/lib/include/libv4l-plugin.h b/lib/include/libv4l-plugin.h
index 23b2dfb..d264529 100644
--- a/lib/include/libv4l-plugin.h
+++ b/lib/include/libv4l-plugin.h
@@ -32,6 +32,8 @@ struct libv4l_dev_ops {
     int (*ioctl)(void *dev_ops_priv, int fd, unsigned long int request, void *arg);
     ssize_t (*read)(void *dev_ops_priv, int fd, void *buffer, size_t n);
     ssize_t (*write)(void *dev_ops_priv, int fd, const void *buffer, size_t n);
+    void * (*mmap)(void *dev_ops_priv, void *start, size_t length, int prot,
+                    int flags, int fd, int64_t offset);
     /* For future plugin API extension, plugins implementing the current API
        must set these all to NULL, as future versions may check for these */
     void (*reserved1)(void);
@@ -40,7 +42,6 @@ struct libv4l_dev_ops {
     void (*reserved4)(void);
     void (*reserved5)(void);
     void (*reserved6)(void);
-    void (*reserved7)(void);
 };

 #endif
diff --git a/lib/libv4l2/libv4l2.c b/lib/libv4l2/libv4l2.c
index 9e4e076..5cb9f3e 100644
--- a/lib/libv4l2/libv4l2.c
+++ b/lib/libv4l2/libv4l2.c
@@ -1642,6 +1642,14 @@ void *v4l2_mmap(void *start, size_t length, int prot, int flags, int fd,
			return MAP_FAILED;
		}

+		if (devices[index].dev_ops->mmap) {
+			result = devices[index].dev_ops->
+					mmap(devices[index].dev_ops_priv, start,
+					     length, prot, flags, fd, offset);
+			if (result != MAP_FAILED)
+				return result;
+		}
+
		return (void *)SYS_MMAP(start, length, prot, flags, fd, offset);
	}

--
1.9.1

